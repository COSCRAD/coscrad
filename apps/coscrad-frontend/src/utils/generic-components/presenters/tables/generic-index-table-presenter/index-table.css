.index-footer {
    border-top-left-radius: 0px !important;
    border-top-right-radius: 0px !important;
    text-align: right;
}

.pagination-control {
    padding-right: 20px;
}

.pagination-arrow {
    vertical-align: sub;
    padding-top: 14px;
}

#pagination-back-arrow {
    padding-left: 30px;
}

#pagination-front-arrow {
    padding-right: 20px;
}
