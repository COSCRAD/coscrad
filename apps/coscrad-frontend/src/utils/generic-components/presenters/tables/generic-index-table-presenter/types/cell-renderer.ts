export type CellRenderer<T> = (row: T) => JSX.Element | string;
