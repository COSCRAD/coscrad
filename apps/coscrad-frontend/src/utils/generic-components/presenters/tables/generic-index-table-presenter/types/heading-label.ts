import { HasId } from '@coscrad/api-interfaces';

export type HeadingLabel<T extends HasId> = {
    propertyKey: keyof T;
    headingLabel: string;
};
