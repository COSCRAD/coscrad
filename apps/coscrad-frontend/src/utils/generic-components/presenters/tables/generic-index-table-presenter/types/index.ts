export * from './cell-renderer';
export * from './cell-renderers-map';
export * from './heading-label';
