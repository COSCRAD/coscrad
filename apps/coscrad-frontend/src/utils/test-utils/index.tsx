export * from './assertions';
export * from './render-with-providers';
