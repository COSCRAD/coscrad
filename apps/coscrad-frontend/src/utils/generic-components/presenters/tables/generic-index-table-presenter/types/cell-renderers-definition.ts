import { CellRenderer } from './cell-renderer';

export type CellRenderersDefinition<T> = {
    [K in keyof T]?: CellRenderer<T>;
};
