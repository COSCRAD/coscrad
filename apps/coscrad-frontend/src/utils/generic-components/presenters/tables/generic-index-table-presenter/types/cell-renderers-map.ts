import { IBaseViewModel } from '@coscrad/api-interfaces';
import { CellRenderer } from './cell-renderer';

export type CellRenderersMap<T extends IBaseViewModel> = Map<keyof T, CellRenderer<T>>;
