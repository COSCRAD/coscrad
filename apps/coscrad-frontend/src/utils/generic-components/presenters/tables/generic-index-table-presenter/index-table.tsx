import { isNullOrUndefined } from '@coscrad/validation-constraints';
import ArrowBackIosIcon from '@mui/icons-material/ArrowBackIos';
import ArrowForwardIosIcon from '@mui/icons-material/ArrowForwardIos';
import {
    Box,
    Checkbox,
    FormControl,
    Grid,
    IconButton,
    InputLabel,
    MenuItem,
    Paper,
    Select,
    Stack,
    Table,
    TableBody,
    TableCell,
    TableContainer as MUITableContainer,
    TableHead,
    TableRow,
    Typography,
} from '@mui/material';
import { useContext, useEffect, useState } from 'react';
import { NotFoundPresenter } from '../../../../../components/not-found';
import { ConfigurableContentContext } from '../../../../../configurable-front-matter/configurable-content-provider';
import { cyclicDecrement, cyclicIncrement } from '../../../../math';
import { EmptyIndexTableException, UnnecessaryCellRendererDefinitionException } from './exceptions';
import { filterTableData, Matchers } from './filter-table-data';
import { renderCell } from './render-cell';
import { SearchBar } from './search-bar';
import { CellRenderer, CellRenderersMap, HeadingLabel } from './types';
import { CellRenderersDefinition } from './types/cell-renderers-definition';

interface HasId {
    id: string;
}

export const DEFAULT_PAGE_SIZE = 5;

const pageSizeOptions: number[] = [DEFAULT_PAGE_SIZE, 10, 50, 100];

const labelForSearchAllPropertiesOption = 'ALL';

const calculateNumberOfPages = (numberOfRecords: number, pageSize: number) => {
    const quotient = Math.floor(numberOfRecords / pageSize);

    const remainder = numberOfRecords % pageSize;

    return remainder === 0 ? quotient : quotient + 1;
};

/**
 * TODO [https://www.pivotaltracker.com/story/show/182694263]
 * Add ValueUnion to the types library.
 */
export type ValueUnion<T> = T[keyof T];

/**
 * We want to constrain the keys of renderers to a subset of the heading
 * labels' property keys. - This could lead to clients specifying unused renderers.
 * For now, we just do a check and throw.
 *
 * We may also want to require renderers for non-string (or maybe non-primitive types)
 */
export interface GenericIndexTablePresenterProps<T> {
    headingLabels: HeadingLabel<T>[];
    tableData: T[];
    cellRenderersDefinition: CellRenderersDefinition<T>;
    heading: string;
    filterableProperties: (keyof T)[];
    matchers?: Matchers<T>;
}

const allProperties = 'allProperties';

export const IndexTable = <T,>({
    headingLabels,
    tableData,
    cellRenderersDefinition,
    heading,
    filterableProperties,
    matchers = {}, // default to String(value) & case-insensitive search
}: GenericIndexTablePresenterProps<T>) => {
    if (headingLabels.length === 0) {
        throw new EmptyIndexTableException();
    }

    // TODO [] Encapsulte this as part of the `SearchBar`.
    const { simulatedKeyboard } = useContext(ConfigurableContentContext);

    const [searchValue, setSearchValue] = useState('');

    // SEARCH LOGIC
    const [selectedFilterProperty, setSelectedFilterProperty] = useState<
        typeof allProperties | keyof T
    >(allProperties);

    const [shouldUseVirtualKeyboard, _setShouldUseVirtualKeyboard] = useState<boolean>(true);

    const propertiesToFilterBy =
        selectedFilterProperty === 'allProperties'
            ? filterableProperties
            : [selectedFilterProperty];

    const filteredTableData = filterTableData(
        tableData,
        propertiesToFilterBy,
        searchValue,
        matchers
    );

    // PAGINATION
    // we index pages starting at 0
    const [currentPageIndex, setCurrentPageIndex] = useState(0);
    const [pageSize, setPageSize] = useState(pageSizeOptions[0]);

    const lastPageIndex = calculateNumberOfPages(filteredTableData.length, pageSize) - 1;

    useEffect(() => {
        if (currentPageIndex > lastPageIndex) setCurrentPageIndex(0);
    }, [lastPageIndex, currentPageIndex]);

    const startIndex = currentPageIndex * pageSize;
    const endIndex = startIndex + pageSize;
    const paginatedData = filteredTableData.slice(startIndex, endIndex);

    /**
     * It's tricky to get type safety that forces cell renderers to only include
     * properties referenced in the heading labels. For now, we'll do a dynamic
     * check instead.
     */
    const propertiesInTable = headingLabels.map(({ propertyKey }) => propertyKey);

    const cellRendererKeysNotInHeadings = Object.keys(cellRenderersDefinition).reduce(
        (acc: string[], rendererPropertyKey) =>
            // @ts-expect-error We need to tell the compiler the keys of IBaseViewModel  must be strings
            propertiesInTable.includes(rendererPropertyKey) ? acc : acc.concat(rendererPropertyKey),
        []
    );

    if (cellRendererKeysNotInHeadings.length > 0) {
        throw new UnnecessaryCellRendererDefinitionException(cellRendererKeysNotInHeadings);
    }

    const cellRenderers: CellRenderersMap<T> = new Map(
        Object.entries(cellRenderersDefinition) as [keyof T, CellRenderer<T>][]
    );

    /**
     * TODO Break the presentation part of this table out so that we can inject
     * instead a mobile list view, for example, without rewriting the filtering
     * and pagination logic.
     */
    const table =
        paginatedData.length === 0 ? (
            <NotFoundPresenter />
        ) : (
            <Box sx={{ width: '100%' }}>
                <Paper>
                    <MUITableContainer>
                        <Table aria-labelledby="Resources Table">
                            <TableHead>
                                <TableRow>
                                    {headingLabels.map(({ headingLabel }) => (
                                        <TableCell sx={{ fontWeight: 'bold' }} key={headingLabel}>
                                            {headingLabel}
                                        </TableCell>
                                    ))}
                                </TableRow>
                            </TableHead>
                            <TableBody>
                                {paginatedData.map((row, index) => (
                                    // TODO find a better fallback key
                                    <TableRow
                                        key={(row as HasId).id || index}
                                        data-testid={(row as HasId).id || index}
                                    >
                                        {headingLabels.map(({ propertyKey }) => (
                                            // A little inversion of control here
                                            // We may want to use some currying here
                                            <TableCell key={String(propertyKey)}>
                                                {renderCell(row, cellRenderers, propertyKey)}
                                            </TableCell>
                                        ))}
                                    </TableRow>
                                ))}
                            </TableBody>
                        </Table>
                    </MUITableContainer>
                    <Box
                        component="div"
                        sx={{
                            display: 'grid',
                            flexDirection: 'row',
                            alignItems: 'center',
                            justifyItems: 'flex-end',
                        }}
                    >
                        <Grid container justifyContent="flex-end" spacing={3}>
                            <Grid item sx={{ display: 'flex', alignItems: 'center' }}>
                                <Typography component="span" sx={{ mr: 2 }}>
                                    Rows per page:
                                </Typography>
                                <FormControl variant="standard" sx={{ m: 1 }}>
                                    <Select
                                        name="pageSize"
                                        value={pageSize}
                                        onChange={(changeEvent) => {
                                            const {
                                                target: { value },
                                            } = changeEvent;

                                            const newPageSize =
                                                typeof value === 'string'
                                                    ? Number.parseInt(value)
                                                    : value;

                                            setPageSize(newPageSize);
                                        }}
                                    >
                                        {pageSizeOptions.map((pageSize) => (
                                            <MenuItem key={pageSize} value={pageSize}>
                                                {pageSize}
                                            </MenuItem>
                                        ))}
                                    </Select>
                                </FormControl>
                            </Grid>
                            <Grid item sx={{ display: 'flex', alignItems: 'center' }}>
                                Page: {currentPageIndex + 1}/{lastPageIndex + 1}
                            </Grid>
                            <Grid item sx={{ display: 'flex', alignItems: 'center' }}>
                                <IconButton
                                    onClick={() =>
                                        setCurrentPageIndex(
                                            cyclicDecrement(currentPageIndex, lastPageIndex + 1)
                                        )
                                    }
                                >
                                    <ArrowBackIosIcon />
                                </IconButton>
                            </Grid>
                            <Grid item sx={{ display: 'flex', alignItems: 'center' }}>
                                <IconButton
                                    onClick={() =>
                                        setCurrentPageIndex(
                                            cyclicIncrement(currentPageIndex, lastPageIndex + 1)
                                        )
                                    }
                                >
                                    <ArrowForwardIosIcon />
                                </IconButton>
                            </Grid>
                        </Grid>
                    </Box>
                </Paper>
            </Box>
        );

    const propertiesToSearchSelectField = (
        <FormControl sx={{ minWidth: 120 }} size={'small'}>
            <InputLabel>Filter</InputLabel>
            <Select
                label={'Filter'}
                value={selectedFilterProperty}
                onChange={(changeEvent) => {
                    const {
                        target: { value },
                    } = changeEvent;
                    setSelectedFilterProperty(value as keyof T);
                }}
            >
                <MenuItem sx={{ minWidth: 120 }} value={'allProperties'}>
                    {labelForSearchAllPropertiesOption}
                </MenuItem>
                {filterableProperties.map((selectedFilterProperty: keyof T & string) => (
                    <MenuItem
                        key={selectedFilterProperty}
                        value={selectedFilterProperty}
                        sx={{ minWidth: 120 }}
                    >
                        {
                            headingLabels.find(
                                ({ propertyKey: labelPropertyKey }) =>
                                    labelPropertyKey === selectedFilterProperty
                            )?.headingLabel
                        }
                    </MenuItem>
                ))}
            </Select>
        </FormControl>
    );

    return (
        <Stack>
            <Typography variant="h2">{heading}</Typography>
            <Box sx={{ display: 'flex', justifyContent: 'center' }}>
                {propertiesToSearchSelectField}

                <SearchBar
                    value={searchValue}
                    onValueChange={setSearchValue}
                    specialCharacterReplacements={
                        shouldUseVirtualKeyboard
                            ? simulatedKeyboard?.specialCharacterReplacements
                            : undefined
                    }
                />
            </Box>
            <Box sx={{ display: 'flex', justifyContent: 'center' }}>
                <Checkbox
                    checked={shouldUseVirtualKeyboard}
                    onChange={() => _setShouldUseVirtualKeyboard(!shouldUseVirtualKeyboard)}
                />

                {!isNullOrUndefined(simulatedKeyboard) && shouldUseVirtualKeyboard ? (
                    <p>Special Character Input Method: {simulatedKeyboard.name}</p>
                ) : (
                    <p>Click to enable input method: {simulatedKeyboard.name}</p>
                )}
            </Box>

            <Box>{table}</Box>
        </Stack>
    );
};
