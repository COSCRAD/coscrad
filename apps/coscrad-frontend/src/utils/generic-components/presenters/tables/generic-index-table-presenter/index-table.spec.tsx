import { HasId } from '@coscrad/api-interfaces';
import { Link, MemoryRouter } from 'react-router-dom';
import { renderWithProviders } from '../../../../test-utils';
import { GenericIndexTablePresenterProps, IndexTable } from './index-table';
import { HeadingLabel } from './types';
import { CellRenderersDefinition } from './types/cell-renderers-definition';

type Widget = HasId & {
    foo: string;
    bar: number[];
    baz: {
        nestedProp: string[];
    };
};

/**
 * We introduce a fake model for the purpose of this test so not to assume
 * anything other than the existence of an `id` property.
 */
const widgets: Widget[] = [
    {
        id: '1',
        foo: 'a',
        bar: [1, 2, 3],
        baz: {
            nestedProp: ['b', 'c'],
        },
    },
    {
        id: '2',
        foo: 'e',
        bar: [11, 233, 34],
        baz: {
            nestedProp: ['c'],
        },
    },
    {
        id: '3',
        foo: 'hello world',
        bar: [1, 23, 34],
        baz: {
            nestedProp: [],
        },
    },
];

const allHeadings: HeadingLabel<Widget>[] = [
    {
        propertyKey: 'id',
        headingLabel: 'LINK',
    },
    {
        propertyKey: 'foo',
        headingLabel: 'Foo',
    },
    {
        propertyKey: 'bar',
        headingLabel: 'The Bars',
    },
    {
        propertyKey: 'baz',
        headingLabel: 'Checkout Baz',
    },
];

const filterableProperties: (keyof Widget)[] = [
    'foo', 'bar', 'baz'
]

const comprehensiveCellRenderersDefinition: CellRenderersDefinition<Widget> = {
    id: ({ id }: Widget) => <Link to={id}>LINK</Link>,
    foo: ({ foo, bar }: Widget) => `combined prop: ${foo} (${bar})`,
    bar: ({ bar }: Widget) => <div>{bar.length > 0 ? bar[0] : 'NO BAR!'}</div>,
    baz: ({ baz: { nestedProp } }: Widget) => nestedProp.join(' and then '),
};

/**
 * If inspecting the snapshot proves tedious or flakey, we may want to check
 * look for a `data-testid` prop on rendered cells.
 */
const assertValidTableRender = (props: GenericIndexTablePresenterProps<Widget>) => {
    // act
    const { baseElement } = renderWithProviders(
        <MemoryRouter>
            <IndexTable {...props} />
        </MemoryRouter>
    );

    expect(baseElement).toMatchSnapshot();
};

const heading = 'Widgets';

describe('IndexTable', () => {
    describe('when the table setup is valid', () => {
        describe('when all properies are included in the table (registered in headings)', () => {
            describe('when there is a cell renderer for every property', () => {
                it('should render properly', () => {
                    assertValidTableRender({
                        headingLabels: allHeadings,
                        tableData: widgets,
                        cellRenderersDefinition: comprehensiveCellRenderersDefinition,
                        heading,
                        filterableProperties: []
                    });
                });
            });

            describe('when a cell renderer is optionally omitted for one property (foo)', () => {
                const partialRenderers = {
                    id: comprehensiveCellRenderersDefinition.id,
                    bar: comprehensiveCellRenderersDefinition.bar,
                    baz: comprehensiveCellRenderersDefinition.baz,
                };

                it('should render properly', () => {
                    assertValidTableRender({
                        headingLabels: allHeadings,
                        tableData: widgets,
                        cellRenderersDefinition: partialRenderers,
                        heading,
                        filterableProperties
                    });
                });
            });

            describe('when all of the cell renders are omitted', () => {
                it('should render properly', () => {
                    assertValidTableRender({
                        headingLabels: allHeadings,
                        tableData: widgets,
                        cellRenderersDefinition: {},
                        heading,
                        filterableProperties
                    });
                });
            });
        });

        describe('when some of the properties are omitted from the table (foo,bar)', () => {
            const partialHeadings = allHeadings.filter(({ propertyKey }) =>
                ['id', 'baz'].includes(propertyKey)
            );

            describe('when there is a cell renderer for every included property', () => {
                const renderers = {
                    id: comprehensiveCellRenderersDefinition.id,
                    baz: comprehensiveCellRenderersDefinition.baz,
                };

                assertValidTableRender({
                    headingLabels: partialHeadings,
                    tableData: widgets,
                    cellRenderersDefinition: renderers,
                    heading,
                    filterableProperties
                });
            });

            describe('when a cell renderer is omitted for an included property (baz)', () => {
                assertValidTableRender({
                    headingLabels: partialHeadings,
                    tableData: widgets,
                    cellRenderersDefinition: {
                        id: comprehensiveCellRenderersDefinition.id,
                    },
                    heading,
                    filterableProperties
                });
            });

            describe('when no cell renderers are specified', () => {
                assertValidTableRender({
                    headingLabels: partialHeadings,
                    tableData: widgets,
                    cellRenderersDefinition: {},
                    heading,
                    filterableProperties
                });
            });
        });
    });

    describe('when the table setup is invalid', () => {
        const partialHeadings = allHeadings.filter(({ propertyKey }) =>
            ['id', 'baz'].includes(propertyKey)
        );
        describe('when there are unnecessary renderers', () => {
            const renderersWithExtra = {
                id: comprehensiveCellRenderersDefinition.id,
                baz: comprehensiveCellRenderersDefinition.baz,
                bar: (_: Widget) => 'I am not necessary- bar is not included in the table',
            };

            const attemptBadRender = () =>
                IndexTable({
                    headingLabels: partialHeadings,
                    tableData: widgets,
                    cellRenderersDefinition: renderersWithExtra,
                    heading,
                    filterableProperties
                });

            it('should throw', () => {
                expect(attemptBadRender).toThrow();
            });
        });

        describe('when there are no headings specified', () => {
            const attemptBadRender = () =>
                IndexTable({
                    headingLabels: [],
                    tableData: widgets,
                    cellRenderersDefinition: {},
                    heading,
                    filterableProperties
                });

            it('should throw', () => {
                expect(attemptBadRender).toThrow();
            });
        });
    });
});
