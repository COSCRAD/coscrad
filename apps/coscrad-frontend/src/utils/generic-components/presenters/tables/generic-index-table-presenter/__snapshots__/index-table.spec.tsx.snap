// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<body>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  b and then c
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  c
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[` 2`] = `
<body>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  b and then c
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  c
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[b,c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[]}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[` 3`] = `
<body>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  b and then c
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  c
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[b,c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[]}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <div>
                    1
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[b,c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <div>
                    2
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <div>
                    3
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[]}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`IndexTable when the table setup is valid when all properies are included in the table (registered in headings) when a cell renderer is optionally omitted for one property (foo) should render properly 1`] = `
<body>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Foo
                </th>
                <th>
                  The Bars
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    a
                  </div>
                </td>
                <td>
                  <div>
                    1
                  </div>
                </td>
                <td>
                  b and then c
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    e
                  </div>
                </td>
                <td>
                  <div>
                    11
                  </div>
                </td>
                <td>
                  c
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    hello world
                  </div>
                </td>
                <td>
                  <div>
                    1
                  </div>
                </td>
                <td />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`IndexTable when the table setup is valid when all properies are included in the table (registered in headings) when all of the cell renders are omitted should render properly 1`] = `
<body>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Foo
                </th>
                <th>
                  The Bars
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <div>
                    1
                  </div>
                </td>
                <td>
                  <div>
                    a
                  </div>
                </td>
                <td>
                  <div>
                    [1,2,3]
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[b,c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <div>
                    2
                  </div>
                </td>
                <td>
                  <div>
                    e
                  </div>
                </td>
                <td>
                  <div>
                    [11,233,34]
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <div>
                    3
                  </div>
                </td>
                <td>
                  <div>
                    hello world
                  </div>
                </td>
                <td>
                  <div>
                    [1,23,34]
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[]}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`IndexTable when the table setup is valid when all properies are included in the table (registered in headings) when there is a cell renderer for every property should render properly 1`] = `
<body>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  b and then c
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  c
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[b,c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  <div>
                    {nestedProp:[]}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <div>
                    1
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[b,c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <div>
                    2
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[c]}
                  </div>
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <div>
                    3
                  </div>
                </td>
                <td>
                  <div>
                    {nestedProp:[]}
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div>
      <h3>
        Widgets
      </h3>
      <div
        class="records-table"
      >
        <div>
          <table>
            <thead>
              <tr>
                <th>
                  LINK
                </th>
                <th>
                  Foo
                </th>
                <th>
                  The Bars
                </th>
                <th>
                  Checkout Baz
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                data-testid="1"
              >
                <td>
                  <a
                    href="/1"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  combined prop: a (1,2,3)
                </td>
                <td>
                  <div>
                    1
                  </div>
                </td>
                <td>
                  b and then c
                </td>
              </tr>
              <tr
                data-testid="2"
              >
                <td>
                  <a
                    href="/2"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  combined prop: e (11,233,34)
                </td>
                <td>
                  <div>
                    11
                  </div>
                </td>
                <td>
                  c
                </td>
              </tr>
              <tr
                data-testid="3"
              >
                <td>
                  <a
                    href="/3"
                  >
                    LINK
                  </a>
                </td>
                <td>
                  combined prop: hello world (1,23,34)
                </td>
                <td>
                  <div>
                    1
                  </div>
                </td>
                <td />
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>
`;
